text/0135-compressed-blob-prefixes.md

```diff
@@ -0,0 +1,101 @@
+
+## Motivation
+
+Currently, a compressed blob does not give any idea of what's inside because the only thing that can be inside, according to the spec, is Wasm. In reality, other blob types are already being used, and more are to come. Blob decompression is a blocking operation, and routing the blob to a proper worker should not involve its decompression for the sake of efficiency. Thus, it is necessary to introduce a mechanism allowing to identify the blob type without decompressing it.
+
+This proposal is intended to:
+1. Fill up gaps in the Polkadot spec, increasing its preciseness and putting it in line with mechanisms already being used in practice but not yet standardized;
+2. Support future work enabling Polkadot to execute PolkaVM and, more generally, other-than-Wasm parachain runtimes;
+3. Allow developers to introduce compression methods other than `zstd` seamlessly in the future.
+
+## Stakeholders
+
+Node developers are the main stakeholders for this proposal. It also creates a foundation on which parachain runtime developers will build.
+
+## Explanation
+
+### Overview
+
+The current approach to compressing binary blobs is defined in [subsection 2.6.2](https://spec.polkadot.network/chap-state#sect-loading-runtime-code) of Polkadot spec. It involves using `zstd` compression, and the resulting compressed blob is prefixed with a unique 64-bit magic value specified in that subsection. Said subsection only defines the means of compressing Wasm code blobs; no other compression procedure is currently defined by the spec.
+
+However, in practice, the current de facto protocol uses the said procedure to compress not only Wasm code blobs but also proofs-of-validity. Such a usage is not stipulated by the spec. Currently, having solely a compressed blob, it's impossible to tell what's inside it without decompression, a Wasm blob, or a PoV. That doesn't cause any problems in the current de facto protocol, as Wasm blobs and PoV blobs take completely different execution paths, and it's impossible to mix them.
+
+Changes proposed below are intended to:
+* Bring the spec into line with the currently used de facto protocol;
+* Define the means for distinguishing compressed blob types in a backward-compatible and future-proof way;
+* Implement the means defined;
+* Deprecate the legacy compression/decompression mechanism in favor of the proposed one.
+
+### Proposed spec changes
+
+* Add a subsection titled "Blob Compression" and describe how the binary data is compressed generally and list the following well-known prefixes:
+
+| Prefix name | Prefix bytes | Description |
+| -- | -- | -- |
+| `CBLOB_ZSTD_WASM_OR_POV` | 82, 188, 83, 118, 70, 219, 142, 5 | Wasm code blob or PoV, zstd-compressed |
+| `CBLOB_ZSTD_WASM_CODE`   | 82, 188, 83, 118, 70, 219, 142, 6 | Wasm code blob, zstd-compressed        |
+| `CBLOB_ZSTD_PVM_CODE`    | 82, 188, 83, 118, 70, 219, 142, 7 | PolkaVM code blob, zstd-compressed     |
```

Comment:
Why don't we just introduce a new variant for `PVM` and say that the old variant is only used for WASM?

Reply:
Because the old variant is currently used for PoVs as well.
Yes, we could leave the old variant as-is and just add PVM to minimize the change footprint. But to me, it looks like a bad design. If we add granularity, let's do it correctly right away, not leaving the technical debt behind. When this table grows, that "wasm-or-pov" thing will look like an obvious legacy and will most probably be harder to remove than now.

Reply:
Now I get what you mean. Then I would say we introduce a third variant that is directly `CBLOB_ZSTD_POV` and doesn't do this dance of renaming it in some months or whatever. Basically deprecating the old prefix.

Reply:
That was my first thought. But the bad news is that the PoVs are compressed by collators. We couldn't remove the old prefix until all the collators upgraded and started using the new one, which could take... Forever

Reply:
We could name it `CBLOB_ZSTD_POV` from the beginning and leave a code comment that it is also used for Wasm blobs, but such usage is deprecated. Removing a comment is a little bit less effort than renaming a constant :upside_down_face:

Reply:
> That was my first thought. But the bad news is that the PoVs are compressed by collators. We couldn't remove the old prefix until all the collators upgraded and started using the new one, which could take... Forever

Yeah and I don't see any real problem with this. We have soo many code that is 5 years old that handles stuff that maybe someone still using somewhere.

Reply:
Personally I agree with Basti. I would probably have a `CBLOB_ZSTD_LEGACY` (or whatever you'd like to call it) which would be the current prefix, and then introduce three new prefixes: one for PVFs, one for WASM, and one for PVM.